packages/core/src/notifications/controller.ts

```diff
@@ -0,0 +1,233 @@
+        this.state.emergency = this.state.emergency ?? {}
+        this.state.dismissed = this.state.dismissed ?? []
+    }
+
+    public pollForStartUp(ruleEngine: RuleEngine) {
+        return this.poll(ruleEngine, 'startUp')
+    }
+
+    public pollForEmergencies(ruleEngine: RuleEngine) {
+        return this.poll(ruleEngine, 'emergency')
+    }
+
+    private async poll(ruleEngine: RuleEngine, category: NotificationType) {
+        try {
+            await this.fetchNotifications(category)
+        } catch (err: any) {
+            getLogger().error(`Unable to fetch %s notifications: %s`, category, err)
+        }
+
+        await this.displayNotifications(ruleEngine)
+    }
+
+    private async displayNotifications(ruleEngine: RuleEngine) {
+        const dismissed = new Set(this.state.dismissed)
+        const startUp =
+            this.state.startUp.payload?.notifications.filter(
+                (n) => !dismissed.has(n.id) && ruleEngine.shouldDisplayNotification(n)
+            ) ?? []
+        const emergency = (this.state.emergency.payload?.notifications ?? []).filter((n) =>
+            ruleEngine.shouldDisplayNotification(n)
+        )
+
+        NotificationsNode.instance.setNotifications(startUp, emergency)
+
+        // Emergency notifications can't be dismissed, but if the user minimizes the panel then
+        // we don't want to focus it each time we set the notification nodes.
+        // So we store it in dismissed once a focus has been fired for it.
+        const newEmergencies = emergency.map((n) => n.id).filter((id) => !dismissed.has(id))
+        if (newEmergencies.length > 0) {
+            this.state.dismissed = [...this.state.dismissed, ...newEmergencies]
+            await this.writeState()
+            void this.notificationsNode.focusPanel()
+        }
+    }
+
+    /**
+     * Permanently hides a notification from view. Only 'startUp' notifications can be dismissed.
+     * Users are able to collapse or hide the notifications panel in native VSC if they want to
+     * hide all notifications.
+     */
+    public async dismissNotification(notificationId: string) {
+        getLogger().debug('Dismissing notification: %s', notificationId)
+        this.state.dismissed.push(notificationId)
+        await this.writeState()
+
+        NotificationsNode.instance.dismissStartUpNotification(notificationId)
+    }
+
+    /**
+     * Fetch notifications from the endpoint and store them in the global state.
+     */
+    private async fetchNotifications(category: NotificationType) {
+        const response = _useLocalFiles ? await this.fetchLocally(category) : await this.fetchRemotely(category)
+        if (!response.content) {
+            getLogger().verbose('No new notifications for category: %s', category)
+            return
+        }
+
+        getLogger().verbose('ETAG has changed for notifications category: %s', category)
+
+        this.state[category].payload = JSON.parse(response.content)
```

Comment:
From experience, either this or the network call can fail due to network issues. Which can result in SyntaxError or some failed request. 

Is there any error handling upstream? I didn't check thoroughly.

Also see if the withTelemetryContext  decorator can help to give the upstream context about any failures that occur in this function. (I have a pending pr for this)

Reply:
No telemetry yet - will be added in a future PR. Will keep this in mind.